### PR TITLE
feat: add OpenCensus tracing

### PR DESCRIPTION
**Pull request by zoercai**
Adds tracing and a SpannerIO parameter that allows all traces to be sampled.

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/build.gradle [hunk 1]
```diff
@@ -132,6 +132,7 @@ dependencies {
   compile library.java.arrow_memory_core
   compile library.java.arrow_vector
 
+  compile "io.opencensus:opencensus-api:0.28.0"
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/build.gradle:135]: Do we use the same version of the API as `runner/google-dataflow`?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java [hunk 2]
```diff
@@ -1395,6 +1405,10 @@ public ReadChangeStream withRpcPriority(RpcPriority rpcPriority) {
       return toBuilder().setRpcPriority(rpcPriority).build();
     }
 
+    public ReadChangeStream withSampleAllTrace(Boolean sampleAllTrace) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1408]: Is it just a yes / no question here, or could I configure the sampling rate I would like?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/DetectNewPartitionsDoFn.java [hunk 3]
```diff
@@ -116,36 +121,42 @@ public ProcessContinuation processElement(
...
-
-    try (ResultSet resultSet = partitionMetadataDao.getPartitionsInState(State.CREATED)) {
-      long currentIndex = tracker.currentRestriction().getFrom();
-
-      while (resultSet.next()) {
-        if (!tracker.tryClaim(currentIndex)) {
-          return ProcessContinuation.stop();
+    try (Scope scope =
+        TRACER
+            .spanBuilder("DetectNewPartitionsDoFn.processElement")
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/DetectNewPartitionsDoFn.java:126]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/ReadChangeStreamPartitionDoFn.java [hunk 4]
```diff
@@ -201,28 +208,42 @@ public ProcessContinuation processElement(
...
-        return deletePartition(partition, tracker);
-      case DONE:
-        return done(partition, tracker);
-      default:
-        // TODO: Verify what to do here
-        LOG.error("[" + token + "] Unknown mode " + mode);
-        throw new IllegalArgumentException("Unknown mode " + mode);
+    try (Scope scope =
+        TRACER
+            .spanBuilder("ReadChangeStreamPartitionDoFn.processElement")
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/ReadChangeStreamPartitionDoFn.java:214]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/ReadChangeStreamPartitionDoFn.java [hunk 5]
```diff
@@ -201,28 +208,42 @@ public ProcessContinuation processElement(
...
+            .setSampler(Samplers.alwaysSample())
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/ReadChangeStreamPartitionDoFn.java:215]: Should we set `Samplers.alwaysSample()` here or should it come from the configuration?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/ChildPartitionsRecordAction.java [hunk 6]
```diff
@@ -55,109 +61,145 @@ public ChildPartitionsRecordAction(PartitionMetadataDao partitionMetadataDao) {
...
-    }
-    watermarkEstimator.setWatermark(new Instant(startTimestamp.toSqlTimestamp().getTime()));
 
-    for (ChildPartition childPartition : record.getChildPartitions()) {
-      if (isSplit(childPartition)) {
-        processSplit(partition, record, childPartition);
-      } else {
-        processMerge(partition, record, childPartition);
+    try (Scope scope =
+        TRACER.spanBuilder("ChildPartitionsRecordAction").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/ChildPartitionsRecordAction.java:66]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/ChildPartitionsRecordAction.java [hunk 7]
```diff
@@ -55,109 +61,145 @@ public ChildPartitionsRecordAction(PartitionMetadataDao partitionMetadataDao) {
...
-                + partitionToken
-                + "] Child partition token "
-                + childPartitionToken
-                + " already exists, skipping");
-      } else {
-        throw e;
+
+    try (Scope scope =
+        TRACER
+            .spanBuilder("ChildPartitionsRecordAction.processSplit")
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/ChildPartitionsRecordAction.java:102]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/ChildPartitionsRecordAction.java [hunk 8]
```diff
@@ -55,109 +61,145 @@ public ChildPartitionsRecordAction(PartitionMetadataDao partitionMetadataDao) {
...
-                + partitionToken
-                + "] Child partition token "
-                + childPartitionToken
-                + " already exists, skipping");
-      } else {
-        throw e;
+
+    try (Scope scope =
+        TRACER
+            .spanBuilder("ChildPartitionsRecordAction.processMerge")
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/ChildPartitionsRecordAction.java:146]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/DataChangeRecordAction.java [hunk 9]
```diff
@@ -34,27 +40,38 @@
...
-      return Optional.of(ProcessContinuation.stop());
-    }
-    // TODO: Ask about this, do we need to output with timestamp?
-    outputReceiver.output(record);
-    watermarkEstimator.setWatermark(new Instant(commitTimestamp.toSqlTimestamp().getTime()));
 
-    LOG.debug("[" + token + "] Data record action completed successfully");
-    return Optional.empty();
+    try (Scope scope =
+        TRACER.spanBuilder("DataChangeRecordAction").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/DataChangeRecordAction.java:53]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/DeletePartitionAction.java [hunk 10]
```diff
@@ -40,16 +48,25 @@ public DeletePartitionAction(PartitionMetadataDao partitionMetadataDao) {
   public Optional<ProcessContinuation> run(
       PartitionMetadata partition,
       RestrictionTracker<PartitionRestriction, PartitionPosition> tracker) {
-    final String token = partition.getPartitionToken();
-    LOG.debug("[" + token + "] Deleting partition");
+    try (Scope scope =
+        TRACER.spanBuilder("DeletePartitionAction").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/DeletePartitionAction.java:52]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/DonePartitionAction.java [hunk 11]
```diff
@@ -28,19 +34,29 @@
...
   private static final Logger LOG = LoggerFactory.getLogger(DonePartitionAction.class);
+  private static final Tracer TRACER = Tracing.getTracer();
 
   public ProcessContinuation run(
       PartitionMetadata partition,
       RestrictionTracker<PartitionRestriction, PartitionPosition> tracker) {
-    final String token = partition.getPartitionToken();
-    LOG.debug("[" + token + "] Marking partition as done");
+    try (Scope scope =
+        TRACER.spanBuilder("DonePartitionAction").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/DonePartitionAction.java:43]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/FinishPartitionAction.java [hunk 12]
```diff
@@ -48,26 +55,35 @@ public FinishPartitionAction(PartitionMetadataDao partitionMetadataDao) {
   public Optional<ProcessContinuation> run(
       PartitionMetadata partition,
       RestrictionTracker<PartitionRestriction, PartitionPosition> tracker) {
-    final String token = partition.getPartitionToken();
-    LOG.debug("[" + token + "] Finishing partition");
+    try (Scope scope =
+        TRACER.spanBuilder("FinishPartitionAction").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/FinishPartitionAction.java:59]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/HeartbeatRecordAction.java [hunk 13]
```diff
@@ -33,23 +39,34 @@
...
-    LOG.debug("[" + token + "] Processing heartbeat record " + record);
 
-    final Timestamp timestamp = record.getTimestamp();
-    if (!tracker.tryClaim(PartitionPosition.queryChangeStream(timestamp))) {
-      LOG.debug("[" + token + "] Could not claim queryChangeStream(" + timestamp + "), stopping");
-      return Optional.of(ProcessContinuation.stop());
-    }
-    watermarkEstimator.setWatermark(new Instant(timestamp.toSqlTimestamp().getTime()));
+    try (Scope scope =
+        TRACER.spanBuilder("HeartbeatRecordAction").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/HeartbeatRecordAction.java:51]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/QueryChangeStreamAction.java [hunk 14]
```diff
@@ -66,61 +74,69 @@ public QueryChangeStreamAction(
       RestrictionTracker<PartitionRestriction, PartitionPosition> tracker,
       OutputReceiver<DataChangeRecord> receiver,
       ManualWatermarkEstimator<Instant> watermarkEstimator) {
-    final String token = partition.getPartitionToken();
+    try (Scope scope =
+        TRACER.spanBuilder("QueryChangeStreamAction").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/QueryChangeStreamAction.java:78]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/QueryChangeStreamAction.java [hunk 15]
```diff
@@ -66,61 +74,69 @@ public QueryChangeStreamAction(
...
+      try (ChangeStreamResultSet resultSet =
+          changeStreamDao.changeStreamQuery(
+              token,
+              tracker.currentRestriction().getStartTimestamp(),
+              partition.isInclusiveStart(),
+              partition.getEndTimestamp(),
+              partition.isInclusiveEnd(),
+              partition.getHeartbeatMillis())) {
+        int count = 0;
+        while (resultSet.next()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/QueryChangeStreamAction.java:95]: Could we trace into one `resultSet.next()` element? Or is this caught by the trace of each underlying action?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/WaitForChildPartitionsAction.java [hunk 16]
```diff
@@ -47,29 +54,42 @@ public WaitForChildPartitionsAction(
...
-    if (numberOfUnscheduledChildren > 0) {
-      LOG.debug(
-          "["
-              + token
-              + " ] Resuming, there are "
-              + numberOfUnscheduledChildren
-              + " unscheduled / not finished children");
+    try (Scope scope =
+        TRACER
+            .spanBuilder("WaitForChildPartitionsAction")
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/WaitForChildPartitionsAction.java:60]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/WaitForParentPartitionsAction.java [hunk 17]
```diff
@@ -44,26 +52,38 @@ public WaitForParentPartitionsAction(
   public Optional<ProcessContinuation> run(
       PartitionMetadata partition,
       RestrictionTracker<PartitionRestriction, PartitionPosition> tracker) {
-    final String token = partition.getPartitionToken();
-    LOG.debug("[" + token + "] Waiting for parent partitions");
+    try (Scope scope =
+        TRACER
+            .spanBuilder("WaitForParentPartitionsAction")
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/WaitForParentPartitionsAction.java:57]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 18]
```diff
@@ -66,70 +73,93 @@ public PartitionMetadataDao(
...
-            .to(partitionToken)
-            .bind("states")
-            .toStringArray(states.stream().map(State::toString).collect(Collectors.toList()))
-            .build();
-    try (ResultSet resultSet = databaseClient.singleUse().executeQuery(statement)) {
-      resultSet.next();
-      return resultSet.getLong(0);
+    try (Scope scope =
+        TRACER
+            .spanBuilder("countChildPartitionsInStates")
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:78]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 19]
```diff
@@ -66,70 +73,93 @@ public PartitionMetadataDao(
...
-                    + "@partition"
-                    + "))")
-            .bind("partition")
-            .to(partitionToken)
-            .build();
-    try (ResultSet resultSet = databaseClient.singleUse().executeQuery(statement)) {
-      resultSet.next();
-      return resultSet.getLong(0);
+    try (Scope scope =
+        TRACER.spanBuilder("countExistingParents").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:110]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 20]
```diff
@@ -66,70 +73,93 @@ public PartitionMetadataDao(
...
     }
   }
 
   public long countPartitions() {
-    final Statement statement = Statement.newBuilder("SELECT COUNT(*) FROM " + tableName).build();
-    try (ResultSet resultSet = databaseClient.singleUse().executeQuery(statement)) {
-      resultSet.next();
-      return resultSet.getLong(0);
+    try (Scope scope =
+        TRACER.spanBuilder("countPartitions").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:144]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 21]
```diff
@@ -66,70 +73,93 @@ public PartitionMetadataDao(
...
 
   public ResultSet getPartitionsInState(State state) {
-    Statement statement =
-        Statement.newBuilder("SELECT * FROM " + tableName + " WHERE State = @state")
-            .bind("state")
-            .to(state.toString())
-            .build();
-    return databaseClient.singleUse().executeQuery(statement);
+    try (Scope scope =
+        TRACER.spanBuilder("getPartitionsInState").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:155]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 22]
```diff
@@ -188,70 +219,115 @@ public InTransactionContext(
     }
 
     public Void insert(PartitionMetadata row) {
-      transaction.buffer(createInsertMutationFrom(row));
-      return null;
+      try (Scope scope = TRACER.spanBuilder("insert").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:222]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 23]
```diff
@@ -188,70 +219,115 @@ public InTransactionContext(
...
+        transaction.buffer(createInsertMutationFrom(row));
+        return null;
+      }
     }
 
     public Void updateToScheduled(String partitionToken) {
-      transaction.buffer(createUpdateMutationFrom(partitionToken, State.SCHEDULED));
-      return null;
+      try (Scope scope =
+          TRACER.spanBuilder("updateToScheduled").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:235]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 24]
```diff
@@ -188,70 +219,115 @@ public InTransactionContext(
...
+        transaction.buffer(createUpdateMutationFrom(partitionToken, State.SCHEDULED));
+        return null;
+      }
     }
 
     public Void updateToRunning(String partitionToken) {
-      transaction.buffer(createUpdateMutationFrom(partitionToken, State.RUNNING));
-      return null;
+      try (Scope scope =
+          TRACER.spanBuilder("updateToRunning").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:247]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 25]
```diff
@@ -188,70 +219,115 @@ public InTransactionContext(
...
+        transaction.buffer(createUpdateMutationFrom(partitionToken, State.RUNNING));
+        return null;
+      }
     }
 
     public Void updateToFinished(String partitionToken) {
-      transaction.buffer(createUpdateMutationFrom(partitionToken, State.FINISHED));
-      return null;
+      try (Scope scope =
+          TRACER.spanBuilder("updateToFinished").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:259]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 26]
```diff
@@ -188,70 +219,115 @@ public InTransactionContext(
...
+                PARTITION_ID_ATTRIBUTE_LABEL, AttributeValue.stringAttributeValue(partitionToken));
+        transaction.buffer(createUpdateMutationFrom(partitionToken, State.FINISHED));
+        return null;
+      }
     }
 
     public Void delete(String partitionToken) {
-      transaction.buffer(createDeleteMutationFrom(partitionToken));
-      return null;
+      try (Scope scope = TRACER.spanBuilder("delete").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:270]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 27]
```diff
@@ -188,70 +219,115 @@ public InTransactionContext(
...
-                          + " IN UNNEST (@states)")
-                  .bind("partitions")
-                  .toStringArray(partitionTokens)
-                  .bind("states")
-                  .toStringArray(states.stream().map(State::toString).collect(Collectors.toList()))
-                  .build())) {
-        resultSet.next();
-        return resultSet.getLong(0);
+      try (Scope scope =
+          TRACER.spanBuilder("countPartitionsInStates").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:283]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java [hunk 28]
```diff
@@ -188,70 +219,115 @@ public InTransactionContext(
...
-                          + " WHERE "
-                          + COLUMN_PARTITION_TOKEN
-                          + " = @partition")
-                  .bind("partition")
-                  .to(partitionToken)
-                  .build())) {
-        if (resultSet.next()) {
-          return mapper.from(resultSet);
+      try (Scope scope =
+          TRACER.spanBuilder("getPartition").setRecordEvents(true).startScopedSpan()) {
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/dao/PartitionMetadataDao.java:310]: nit: can we extract a constant for the String?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java (hunk 2 above)
Comment (thiagotnunes) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1408]: Perhaps exposing the `Samplers.method` here is more generic?

File: sdks/java/io/google-cloud-platform/build.gradle (hunk 1 above)
Comment (zoercai) [reply sdks/java/io/google-cloud-platform/build.gradle:135]: yep!

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/ReadChangeStreamPartitionDoFn.java (hunk 5 above)
Comment (zoercai) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/ReadChangeStreamPartitionDoFn.java:215]: ah missed this one when I was refactoring, thanks for catching it!

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/QueryChangeStreamAction.java (hunk 15 above)
Comment (zoercai) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/cdc/actions/QueryChangeStreamAction.java:95]: Yep I expect here the underlying actions will be nested under each element

**Review by thiagotnunes — APPROVED** [review]

**Pull request merged**